### PR TITLE
Add kaHIP k-way partitioning

### PR DESCRIPTION
**Pull request by mgsium**

### CODE REVIEW

File: makefiles/dependencies/kahip.mk [hunk 1]
```diff
@@ -0,0 +1,37 @@
+KAHIP_DEF ?= -DHAVE_KAHIP
+
+ifdef KAHIP_INSTALL_PATH
+  KAHIP_INC_PATH := -I$(KAHIP_INSTALL_PATH)/parallel/parallel_src/interface \
```
Comment (reguly) [inline makefiles/dependencies/kahip.mk:4]: This should just be `-I$(KAHIP_INSTALL_PATH)/include`
if the kahip library is properly "installed" i.e. during compilation you have `-DCMAKE_INSTALL_PREFIX=somedir` then during OP2 make config you have `KAHIP_INSTALL_PATH=somedir`

File: makefiles/dependencies/kahip.mk [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+                    -I$(KAHIP_INSTALL_PATH)/interface
+  KAHIP_LIB_PATH := -L$(KAHIP_INSTALL_PATH)/build/parallel/parallel_src \
```
Comment (reguly) [inline makefiles/dependencies/kahip.mk:6]: Same here

File: op2/src/mpi/op_mpi_part_core.cpp [hunk 3]
```diff
@@ -68,6 +68,14 @@ typedef int idx_t;
 #endif
 #endif
 
+// kaHIP header
+#ifdef HAVE_KAHIP
+#include <parhip_interface.h>
+
+typedef idxtype idx_t;
```
Comment (reguly) [inline op2/src/mpi/op_mpi_part_core.cpp:75]: I am getting a compile time error with gcc 9 if I also have ParMetis installed:

> In file included from src/mpi/op_mpi_part_core.cpp:73:
/home/shared/software/kahip-ompi401/include/parhip_interface.h:15:28: error: conflicting declaration ‘typedef long long unsigned int idxtype’
   15 | typedef unsigned long long idxtype;
      |                            ^~~~~~~
src/mpi/op_mpi_part_core.cpp:65:15: note: previous declaration as ‘typedef idx_t idxtype’
   65 | typedef idx_t idxtype;

**Review by reguly — CHANGES_REQUESTED** [review]
I have some compilation issues when I have Parmetis and kahip. Goes through OK if I only have KaHIP.

File: op2/src/mpi/op_mpi_part_core.cpp (hunk 3 above)
Comment (mgsium) [reply op2/src/mpi/op_mpi_part_core.cpp:75]: I ran into the same issue yesterday, looks like ParMETIS defines `idx_t` while ParHIP defines `idxtype`, which is distinct. As long as only one of the two are installed, we can use `typedef` to set the type as necessary but with both installed we're forced to choose between the types at runtime. 

The interface for the two k-way partition functions are very similar apart from this type issue so I'm reluctant to define another `op_partition_kway` just for the ParHIP version. But I've yet to find an elegant solution.

Comment (reguly) [reply op2/src/mpi/op_mpi_part_core.cpp:75]: That's a tough one... My first idea, while being wasteful, but at least compact, would be to pass the parmetis-style arrays to a function, and allocate new ones with kahip types, convert them, then call kahip. Which is indeed ugly! But the fundamental issue is that even if both use 64 bits, kahip uses unsigned, whereas parmetis signed...

**Comment by mgsium** [comment]
I think some overloading is still necessary because the two partition functions don't take the same arguments. The best I could think of was:

```cpp
/*******************************************************************************
 * Generic wrapper for kway-partition functions
 *******************************************************************************/

template <class T>
void do_partition(T *vtxdist, T *xadj, T *adjncy, T *wgtflag, T *numflag, 
            T *ncon, T *nparts,real_t *tpwgts, real_t *ubvec, T *options, 
            T *edgecut, T *part, MPI_Comm *comm) {}

#ifdef HAVE_PARMETIS
template<> void do_partition<idx_t>(idx_t *vtxdist, idx_t *xadj, idx_t *adjncy, 
            idx_t *wgtflag, idx_t *numflag, idx_t *ncon, idx_t *nparts,real_t *tpwgts, 
            real_t *ubvec, idx_t *options, idx_t *edgecut, idx_t *part, MPI_Comm *comm) {
  ParMETIS_V3_PartKway( vtxdist, xadj, adjncy, NULL, NULL, wgtflag, numflag, ncon, 
                        nparts, tpwgts, ubvec, options, edgecut, part, comm);
}
#endif

#ifdef HAVE_KAHIP
template<> void do_partition<idxtype>(idxtype *vtxdist, idxtype *xadj, idxtype *adjncy, 
            idxtype *wgtflag, idxtype *numflag, idxtype *ncon, idxtype *nparts, real_t *tpwgts, 
            real_t *ubvec, idxtype *options, idxtype *edgecut, idxtype *part, MPI_Comm *comm) {
  double imb = 0.03;
  ParHIPPartitionKWay( vtxdist, xadj, adjncy, NULL, NULL, (int*) nparts, &imb, 
                       false, 1, ULTRAFASTMESH, (int*) edgecut, part, comm);
}
#endif
```

The long argument list is a little ugly (although it can be made a little better by omitting the names of unused arguments).
Not an ideal solution but it feels less cumbersome than overloading `op_partition_kway`.

**Comment by reguly** [comment]
Could you not template the whole op_partition_kway function?

i.e.
template <bool use_kahip, typename indextype>
void op_partition_kway() {
then use indextype throughout
}

**Comment by mgsium** [comment]
I'm doing that too, but there's still the problem of choosing whether to call the kahip or parmetis function at runtime.

**Comment by mgsium** [comment]
On second thoughts, `do_partition` doesn't need to be templated. This should suffice:
```cpp
#ifdef HAVE_PARMETIS
void do_partition(idx_t *vtxdist, idx_t *xadj, ...) {
  ParMETIS_V3_PartKway( vtxdist, xadj, adjncy, NULL, NULL, wgtflag, numflag, ncon, 
                        nparts, tpwgts, ubvec, options, edgecut, part, comm);
}
#endif

#ifdef HAVE_KAHIP
void do_partition(idxtype *vtxdist, idxtype *xadj, ...) {
  double imb = 0.03;
  ParHIPPartitionKWay( vtxdist, xadj, adjncy, NULL, NULL, (int*) nparts, &imb, 
                       false, 1, ULTRAFASTMESH, (int*) edgecut, part, comm);
}
#endif
```

**Comment by reguly** [comment]
let's see :)

**Review by reguly — APPROVED** [review]
Compiles and works!

**Comment by mgsium** [comment]
Great! I'll just change the include and library paths and that should be it

**Comment by gihanmudalige** [comment]
@reguly  thanks for following up on this. Seems @mgsium has completed the work in record time !

@mgsium I assume this is now working with the Makefile system in the master branch ?
If we are to merge this I think it would be good to add some documentation on the new capabilities.  For example adding details to  : https://op2-dsl.readthedocs.io/en/latest/api.html?highlight=partition#c.op_partition 
Furthermore if there is a need to have any runtime flags to activate this partiotioner  capabilities, that needs to be documented as well. Finally have you checked the coding standards and run it through git clang-format ?

Thanks.

**Comment by bozbez** [comment]
The Makefiles look good to me.
Once this is merged it would also be worth updating the [spack recipe](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/op2-dsl/package.py).

**Comment by mgsium** [comment]
@gihanmudalige I updated the docs and ran clang-format on the changes. I've been using the Makefile to build and test so all should be good there.

**Pull request merged**